File: sandbox.register/src/kz/greetgo/sandbox/register/util/my_batis/CustomEnumTypeHandler.java
///MODIFY replace sandbox {PROJECT_NAME}
package kz.greetgo.sandbox.register.util.my_batis;

import org.apache.ibatis.type.BaseTypeHandler;
import org.apache.ibatis.type.JdbcType;

import java.sql.CallableStatement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class CustomEnumTypeHandler<E extends Enum<E>> extends BaseTypeHandler<E> {

  private Class<E> type;

  public CustomEnumTypeHandler(Class<E> type) {
    if (type == null) {
      throw new IllegalArgumentException("Type argument cannot be null");
    }
    this.type = type;
  }

  @Override
  public void setNonNullParameter(PreparedStatement ps, int i, E parameter, JdbcType jdbcType) throws SQLException {
    if (jdbcType == null) {
      ps.setString(i, parameter.name());
    } else {
      ps.setObject(i, parameter.name(), jdbcType.TYPE_CODE); // see r3589
    }
  }

  @Override
  public E getNullableResult(ResultSet rs, String columnName) throws SQLException {
    return valueOfOrNull(rs.getString(columnName));
  }

  @Override
  public E getNullableResult(ResultSet rs, int columnIndex) throws SQLException {
    return valueOfOrNull(rs.getString(columnIndex));
  }

  @Override
  public E getNullableResult(CallableStatement cs, int columnIndex) throws SQLException {
    return valueOfOrNull(cs.getString(columnIndex));
  }

  private E valueOfOrNull(String str) {

    if (str == null) {
      return null;
    }

    try {
      return Enum.valueOf(type, str);
    } catch (IllegalArgumentException e) {
      return null;
    }

  }
}

File: sandbox.register/src/kz/greetgo/sandbox/register/util/my_batis/CustomEnumTypeHandler.kt
///MODIFY replace sandbox {PROJECT_NAME}
package kz.greetgo.sandbox.register.util.my_batis

import org.apache.ibatis.type.BaseTypeHandler
import org.apache.ibatis.type.JdbcType

import java.sql.CallableStatement
import java.sql.PreparedStatement
import java.sql.ResultSet

class CustomEnumTypeHandler<E : Enum<E>>(private val type: Class<E>?) : BaseTypeHandler<E>() {

  init {
    if (type == null) {
      throw IllegalArgumentException("Type argument cannot be null")
    }
  }

  override fun setNonNullParameter(ps: PreparedStatement, i: Int, parameter: E, jdbcType: JdbcType?) {
    if (jdbcType == null) {
      ps.setString(i, parameter.name)
    } else {
      ps.setObject(i, parameter.name, jdbcType.TYPE_CODE) // see r3589
    }
  }

  override fun getNullableResult(rs: ResultSet, columnName: String): E? {
    return valueOfOrNull(rs.getString(columnName))
  }

  override fun getNullableResult(rs: ResultSet, columnIndex: Int): E? {
    return valueOfOrNull(rs.getString(columnIndex))
  }

  override fun getNullableResult(cs: CallableStatement, columnIndex: Int): E? {
    return valueOfOrNull(cs.getString(columnIndex))
  }

  private fun valueOfOrNull(str: String?): E? {

    if (str == null) {
      return null
    }

    try {
      return java.lang.Enum.valueOf(type, str)
    } catch (e: IllegalArgumentException) {
      return null
    }

  }
}